File: ios/OpenTokReactNative/Utils/EventUtils.swift
//
//  EventUtils.swift
//  OpenTokReactNative
//
//  Created by Manik Sachdeva on 11/3/18.
//  Copyright © 2018 TokBox Inc. All rights reserved.
//

import Foundation

class EventUtils {
    
    static var sessionPreface: String = "session:";
    static var publisherPreface: String = "publisher:";
    static var subscriberPreface: String = "subscriber:";
    
    static func prepareJSConnectionEventData(_ connection: OTConnection) -> Dictionary<String, Any> {
        var connectionInfo: Dictionary<String, Any> = [:];
        guard connection != nil else { return connectionInfo }
        connectionInfo["connectionId"] = connection.connectionId;
        connectionInfo["creationTime"] = convertDateToString(connection.creationTime);
        connectionInfo["data"] = connection.data;
        return connectionInfo;
    }
    
    static func prepareJSStreamEventData(_ stream: OTStream) -> Dictionary<String, Any> {
        var streamInfo: Dictionary<String, Any> = [:];
        guard OTRN.sharedState.sessions[stream.session.sessionId] != nil else { return streamInfo }
        streamInfo["streamId"] = stream.streamId;
        streamInfo["name"] = stream.name;
        streamInfo["connectionId"] = stream.connection.connectionId;
        streamInfo["connection"] = prepareJSConnectionEventData(stream.connection);
        streamInfo["hasAudio"] = stream.hasAudio;
        streamInfo["sessionId"] = stream.session.sessionId;
        streamInfo["hasVideo"] = stream.hasVideo;
        streamInfo["creationTime"] = convertDateToString(stream.creationTime);
        streamInfo["height"] = stream.videoDimensions.height;
        streamInfo["width"] = stream.videoDimensions.width;
        streamInfo["videoType"] = stream.videoType == OTStreamVideoType.screen ? "screen" : "camera"
        return streamInfo;
    }
    
    static func prepareJSErrorEventData(_ error: OTError) -> Dictionary<String, Any> {
        var errorInfo: Dictionary<String, Any> = [:];
        errorInfo["code"] = error.code;
        errorInfo["message"] = error.localizedDescription;
        return errorInfo;
    }
    
    static func prepareStreamPropertyChangedEventData(_ changedProperty: String, oldValue: Any, newValue: Any, stream: Dictionary<String, Any>) -> Dictionary<String, Any> {
        var streamPropertyEventData: Dictionary<String, Any> = [:];
        streamPropertyEventData["oldValue"] = oldValue;
        streamPropertyEventData["newValue"] = newValue;
        streamPropertyEventData["stream"] = stream;
        streamPropertyEventData["changedProperty"] = changedProperty;
        return streamPropertyEventData
    }

    static func preparePublisherRtcStats(_ rtcStatsReport: [OTPublisherRtcStats]) -> [Dictionary<String, Any>] {
        var statsArray:[Dictionary<String, Any>] = [];
        for value in rtcStatsReport {
            var stats:Dictionary<String, Any> = [:];
            stats["connectionId"] = value.connectionId;
            stats["jsonArrayOfReports"] = value.jsonArrayOfReports;
            statsArray.append(stats);
        }
        return statsArray;
    }

    static func preparePublisherVideoNetworkStats(_ videoStats: [OTPublisherKitVideoNetworkStats]) -> [Dictionary<String, Any>] {
        var statsArray:[Dictionary<String, Any>] = [];
        for value in videoStats {
            var stats: Dictionary<String, Any> = [:];
            stats["connectionId"] = value.connectionId;
            stats["subscriberId"] = value.subscriberId;
            stats["videoPacketsLost"] = value.videoPacketsLost;
            stats["videoBytesSent"] = value.videoBytesSent;
            stats["videoPacketsSent"] = value.videoPacketsSent;
            stats["timestamp"] = value.timestamp;
            statsArray.append(stats);
        }
        return statsArray;
    }

    static func preparePublisherAudioNetworkStats(_ audioStats: [OTPublisherKitAudioNetworkStats]) -> [Dictionary<String, Any>] {
        var statsArray:[Dictionary<String, Any>] = [];
        for value in audioStats {
            var stats: Dictionary<String, Any> = [:];
            stats["connectionId"] = value.connectionId;
            stats["subscriberId"] = value.subscriberId;
            stats["audioPacketsLost"] = value.audioPacketsLost;
            stats["audioPacketsSent"] = value.audioPacketsSent;
            stats["audioBytesSent"] = value.audioBytesSent;
            stats["timestamp"] = value.timestamp;
            statsArray.append(stats);
        }
        return statsArray;
    }
    
    static func prepareSubscriberVideoNetworkStatsEventData(_ videoStats: OTSubscriberKitVideoNetworkStats) -> Dictionary<String, Any> {
        var videoStatsEventData: Dictionary<String, Any> = [:];
        videoStatsEventData["videoPacketsLost"] = videoStats.videoPacketsLost;
        videoStatsEventData["videoBytesReceived"] = videoStats.videoBytesReceived;
        videoStatsEventData["videoPacketsReceived"] = videoStats.videoPacketsReceived;
        videoStatsEventData["timestamp"] = videoStats.timestamp;
        return videoStatsEventData;
    }
    
    static func prepareSubscriberAudioNetworkStatsEventData(_ audioStats: OTSubscriberKitAudioNetworkStats) -> Dictionary<String, Any> {
        var audioStatsEventData: Dictionary<String, Any> = [:];
        audioStatsEventData["audioPacketsLost"] = audioStats.audioPacketsLost;
        audioStatsEventData["audioBytesReceived"] = audioStats.audioBytesReceived;
        audioStatsEventData["audioPacketsReceived"] = audioStats.audioPacketsReceived;
        audioStatsEventData["timestamp"] = audioStats.timestamp;
        return audioStatsEventData;
    }
    
    static func prepareJSSessionEventData(_ session: OTSession) -> Dictionary<String, Any> {
        var sessionInfo: Dictionary<String, Any> = [:];
        sessionInfo["sessionId"] = session.sessionId;
        guard let connection = session.connection else { return sessionInfo };
        sessionInfo["connection"] = prepareJSConnectionEventData(connection);
        return sessionInfo;
    }
    
    static func getSupportedEvents() -> [String] {
        return [
            "\(sessionPreface)streamCreated",
            "\(sessionPreface)streamDestroyed",
            "\(sessionPreface)sessionDidConnect",
            "\(sessionPreface)sessionDidDisconnect",
            "\(sessionPreface)connectionCreated",
            "\(sessionPreface)connectionDestroyed",
            "\(sessionPreface)didFailWithError",
            "\(publisherPreface)streamCreated",
            "\(sessionPreface)signal",
            "\(sessionPreface)muteForced",
            "\(publisherPreface)streamDestroyed",
            "\(publisherPreface)didFailWithError",
            "\(publisherPreface)audioLevelUpdated",
            "\(publisherPreface)rtcStatsReport",
            "\(publisherPreface)muteForced",
            "\(publisherPreface)videoEnabled",
            "\(publisherPreface)videoDisabled",
            "\(publisherPreface)videoDisableWarning",
            "\(publisherPreface)videoDisableWarningLifted",
            "\(subscriberPreface)subscriberDidConnect",
            "\(subscriberPreface)subscriberDidDisconnect",
            "\(subscriberPreface)didFailWithError",
            "\(subscriberPreface)videoNetworkStatsUpdated",
            "\(subscriberPreface)audioNetworkStatsUpdated",
            "\(subscriberPreface)audioLevelUpdated",
            "\(subscriberPreface)rtcStatsReport",
            "\(subscriberPreface)subscriberVideoEnabled",
            "\(subscriberPreface)subscriberVideoDisabled",
            "\(subscriberPreface)subscriberVideoDisableWarning",
            "\(subscriberPreface)subscriberVideoDisableWarningLifted",
            "\(subscriberPreface)subscriberVideoDataReceived",
            "\(sessionPreface)archiveStartedWithId",
            "\(sessionPreface)archiveStoppedWithId",
            "\(sessionPreface)sessionDidBeginReconnecting",
            "\(sessionPreface)sessionDidReconnect",
            "\(sessionPreface)streamPropertyChanged",
            "\(subscriberPreface)subscriberDidReconnect",
            "\(subscriberPreface)subscriberCaptionReceived"
        ];
    }
    
    static func convertDateToString(_ creationTime: Date) -> String {
        let dateFormatter: DateFormatter = DateFormatter();
        dateFormatter.dateFormat = "yyyy-MM-dd HH:mm:ss";
        dateFormatter.timeZone = TimeZone(abbreviation: "UTC");
        return dateFormatter.string(from:creationTime);
    }

    static func createErrorMessage(_ message: String) -> Dictionary<String, String> {
        var errorInfo: Dictionary<String, String> = [:]
        errorInfo["message"] = message
        return errorInfo
    }
    
}

File: android/src/main/java/com/opentokreactnative/utils/EventUtils.java
package com.opentokreactnative.utils;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.WritableArray;
import com.facebook.react.bridge.WritableMap;
import com.opentok.android.Connection;
import com.opentok.android.MediaUtils;
import com.opentok.android.OpentokError;
import com.opentok.android.Session;
import com.opentok.android.Stream;
import com.opentok.android.SubscriberKit;
import com.opentok.android.PublisherKit;

public final class EventUtils {

    public static WritableMap prepareJSConnectionMap(Connection connection) {

        WritableMap connectionInfo = Arguments.createMap();
        if (connection != null) {
            connectionInfo.putString("connectionId", connection.getConnectionId());
            connectionInfo.putString("creationTime", connection.getCreationTime().toString());
            connectionInfo.putString("data", connection.getData());
        }
        return connectionInfo;
    }

    public static WritableMap prepareJSStreamMap(Stream stream, Session session) {

        WritableMap streamInfo = Arguments.createMap();
        if (stream != null) {
            streamInfo.putString("streamId", stream.getStreamId());
            streamInfo.putInt("height", stream.getVideoHeight());
            streamInfo.putInt("width", stream.getVideoWidth());
            streamInfo.putString("creationTime", stream.getCreationTime().toString());
            streamInfo.putString("connectionId", stream.getConnection().getConnectionId());
            streamInfo.putString("sessionId", session.getSessionId());
            streamInfo.putMap("connection", prepareJSConnectionMap(stream.getConnection()));
            streamInfo.putString("name", stream.getName());
            streamInfo.putBoolean("hasAudio", stream.hasAudio());
            streamInfo.putBoolean("hasVideo", stream.hasVideo());
            if (stream.getStreamVideoType().equals(Stream.StreamVideoType.StreamVideoTypeScreen)) {
                streamInfo.putString("videoType", "screen");
            } else {
                streamInfo.putString("videoType", "camera");
            }
        }
        return streamInfo;
    }

    public static WritableMap prepareJSErrorMap(OpentokError error) {

        WritableMap errorInfo = Arguments.createMap();
        errorInfo.putString("message", error.getMessage());
        errorInfo.putString("code", error.getErrorCode().toString());
        return errorInfo;
    }

    public static WritableMap prepareJSSessionMap(Session session) {

        WritableMap sessionInfo = Arguments.createMap();
        sessionInfo.putString("sessionId", session.getSessionId());
        if (session.getConnection() != null) {
            WritableMap connectionInfo = prepareJSConnectionMap(session.getConnection());
            sessionInfo.putMap("connection", connectionInfo);
        }
        return sessionInfo;
    }

    public static WritableMap prepareStreamPropertyChangedEventData(String changedProperty, String oldValue, String newValue, Stream stream, Session session) {

        WritableMap streamPropertyEventData = Arguments.createMap();
        streamPropertyEventData.putString("changedProperty", changedProperty);
        streamPropertyEventData.putString("oldValue", oldValue);
        streamPropertyEventData.putString("newValue", newValue);
        streamPropertyEventData.putMap("stream", prepareJSStreamMap(stream, session));
        return streamPropertyEventData;
    }

    public static WritableMap prepareStreamPropertyChangedEventData(String changedProperty, WritableMap oldValue, WritableMap newValue, Stream stream, Session session) {

        WritableMap streamPropertyEventData = Arguments.createMap();
        streamPropertyEventData.putString("changedProperty", changedProperty);
        streamPropertyEventData.putMap("oldValue", oldValue);
        streamPropertyEventData.putMap("newValue", newValue);
        streamPropertyEventData.putMap("stream", prepareJSStreamMap(stream, session));
        return streamPropertyEventData;
    }

    public static WritableMap prepareStreamPropertyChangedEventData(String changedProperty, Boolean oldValue, Boolean newValue, Stream stream, Session session) {

        WritableMap streamPropertyEventData = Arguments.createMap();
        streamPropertyEventData.putString("changedProperty", changedProperty);
        streamPropertyEventData.putBoolean("oldValue", oldValue);
        streamPropertyEventData.putBoolean("newValue", newValue);
        streamPropertyEventData.putMap("stream", prepareJSStreamMap(stream, session));
        return streamPropertyEventData;
    }

    public static WritableMap prepareAudioNetworkStats(SubscriberKit.SubscriberAudioStats stats) {

        WritableMap audioStats = Arguments.createMap();
        audioStats.putInt("audioPacketsLost", stats.audioPacketsLost);
        audioStats.putInt("audioBytesReceived", stats.audioBytesReceived);
        audioStats.putInt("audioPacketsReceived", stats.audioPacketsReceived);
        audioStats.putDouble("timestamp", stats.timeStamp);
        return audioStats;
    }

    public static WritableMap prepareVideoNetworkStats(SubscriberKit.SubscriberVideoStats stats) {

        WritableMap videoStats = Arguments.createMap();
        videoStats.putInt("videoPacketsLost", stats.videoPacketsLost);
        videoStats.putInt("videoBytesReceived", stats.videoBytesReceived);
        videoStats.putInt("videoPacketsReceived", stats.videoPacketsReceived);
        videoStats.putDouble("timestamp", stats.timeStamp);
        return videoStats;
    }

    public static WritableArray preparePublisherRtcStats(PublisherKit.PublisherRtcStats[] stats) {
        WritableArray statsArrayMap = Arguments.createArray();
        for (PublisherKit.PublisherRtcStats stat : stats) {
          WritableMap statMap = Arguments.createMap();
          statMap.putString("connectionId", stat.connectionId);
          statMap.putString("jsonArrayOfReports", stat.jsonArrayOfReports);
          statsArrayMap.pushMap(statMap);
        }
        return statsArrayMap;
    }

    public static WritableArray preparePublisherAudioStats(PublisherKit.PublisherAudioStats[] stats) {
        WritableArray statsArrayMap = Arguments.createArray();
        for (PublisherKit.PublisherAudioStats stat : stats) {
          WritableMap audioStats = Arguments.createMap();
          audioStats.putString("connectionId", stat.connectionId);
          audioStats.putString("subscriberId", stat.subscriberId);
          audioStats.putDouble("audioBytesSent", stat.audioBytesSent);
          audioStats.putDouble("audioPacketsLost", stat.audioPacketsLost);
          audioStats.putDouble("audioPacketsSent", stat.audioPacketsSent);
          audioStats.putDouble("startTime", stat.startTime);
          statsArrayMap.pushMap(audioStats);
        }
        return statsArrayMap;
    }

    public static WritableArray preparePublisherVideoStats(PublisherKit.PublisherVideoStats[] stats) {
        WritableArray statsArrayMap = Arguments.createArray();
        for (PublisherKit.PublisherVideoStats stat : stats) {
          WritableMap videoStats = Arguments.createMap();
          videoStats.putString("connectionId", stat.connectionId);
          videoStats.putString("subscriberId", stat.subscriberId);
          videoStats.putDouble("videoBytesSent", stat.videoBytesSent);
          videoStats.putDouble("videoPacketsLost", stat.videoPacketsLost);
          videoStats.putDouble("videoPacketsSent", stat.videoPacketsSent);
          videoStats.putDouble("startTime", stat.startTime);
          statsArrayMap.pushMap(videoStats);
        }
        return statsArrayMap;
    }

    public static WritableMap prepareMediaCodecsMap(MediaUtils.SupportedCodecs supportedCodecs) {
        WritableMap codecsMap = Arguments.createMap();
        WritableArray videoDecoderCodecsArray = Arguments.createArray();
        WritableArray videoEncoderCodecsArray = Arguments.createArray();
        for (MediaUtils.VideoCodecType decoderCodec : supportedCodecs.videoDecoderCodecs ) {
            if (decoderCodec.equals(MediaUtils.VideoCodecType.VIDEO_CODEC_H264)) {
                videoDecoderCodecsArray.pushString("H.264");
            } else {
                videoDecoderCodecsArray.pushString("VP8");
            }
        }
        for (MediaUtils.VideoCodecType encoderCodec : supportedCodecs.videoEncoderCodecs ) {
            if (encoderCodec.equals(MediaUtils.VideoCodecType.VIDEO_CODEC_H264)) {
                videoEncoderCodecsArray.pushString("H.264");
            } else {
                videoEncoderCodecsArray.pushString("VP8");
            }
        }
        codecsMap.putArray("videoDecoderCodecs", videoDecoderCodecsArray);
        codecsMap.putArray("videoEncoderCodecs", videoEncoderCodecsArray);
        return codecsMap;
    }

    public static WritableMap createError(String message) {

        WritableMap errorInfo = Arguments.createMap();
        errorInfo.putString("message", message);
        return errorInfo;
    }
}
